### PR TITLE
GEODE-5501: Wait for listener to disconnect

### PR DESCRIPTION
**Pull request by dhemery**
In several tests, a listener calls disconnectFromDS() to disconnect its member when a certain event occurs.

The tests do not wait for the listener to finish disconnecting.

When the tests end, the cleanup also calls `disconnectFromDS()` to disconnect the same member.

Within `disconnectFromDS()`, `InternalDistributedSystem.disconnect()` is called twice, once on the system the test framework knows about, and once on whatever system `InternalDistributedSystem.getConnectedInstance()` returns.

If two calls to `disconnect()` are in progress at the same time, one blocks until the other finishes.

If the test ends before the listener finishes, the listener appears to wait until a later test reuses the same VM, at which time it disconnects whatever system it finds there. We do not know why it waits. We suspect that multiple calls to `disconnect()` (from the listener and from the framework's normal cleanup routine) are interacting in some way.

The later test fails if the listener's second call to `disconnect()` occurs after the test has started a member in the same VM.

To prevent the problem, we:
* Wait for the listener to finish before ending the test.
* Remove the listener at the end of the test, so that it does not
respond to events in later tests.

Signed-Off-By: Helena Bales <hbales@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by dhemery** [comment]
The failures in the DistributedTest run (https://concourse.apachegeode-ci.info/builds/3683) are unrelated.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/NetSearchMessagingDUnitTest.java [hunk 1]
```diff
@@ -216,91 +223,70 @@ public void testNetSearchFailoverFromReplicate() {
...
-    long vm1Count = getReceivedMessages(vm1);
-    long vm2Count = getReceivedMessages(vm2);
-    long vm3Count = getReceivedMessages(vm3);
-
     assertEquals("b", get(vm3, "a"));
 
-    // Make sure we were disconnected in vm0
-    vm0.invoke(new SerializableRunnable("check disconnected") {
+    boolean disconnected = false;
+    while (!disconnected) {
```
Comment (upthewaterspout) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/NetSearchMessagingDUnitTest.java]: Can we make this an Awaitility call?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/NetSearchMessagingDUnitTest.java [hunk 2]
```diff
@@ -419,4 +405,36 @@ public void run() {
...
+      public void run() {
+        listenerHasFinished.set(false);
+        DistributionMessageObserver ob = new DistributionMessageObserver() {
+          public void beforeProcessMessage(ClusterDistributionManager dm,
+              DistributionMessage message) {
+            if (message instanceof NetSearchRequestMessage) {
+              DistributionMessageObserver.setInstance(null);
+              assertThat(getTestMethodName())
+                  .describedAs("test method when listener invoked")
+                  .isEqualTo(installerMethodName);
```
Comment (upthewaterspout) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/NetSearchMessagingDUnitTest.java]: Do we need to make assertions about the test method name here?

**Comment by dhemery** [comment]
Rebased to pick up GeodeAwaitility.

**Pull request merged**